src/dodona/util/Interactive.java

```diff
@@ -0,0 +1,189 @@
+package dodona.util;
+
+import dodona.junit.MessageWriter;
+import org.junit.Assert;
+import org.junit.contrib.java.lang.system.SystemOutRule;
+import org.junit.contrib.java.lang.system.TextFromStandardInputStream;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static org.junit.contrib.java.lang.system.TextFromStandardInputStream.emptyStandardInputStream;
+
+/**
+ * Allows an interactive application to be tested conveniently.
+ */
+public class Interactive implements TestRule {
+    private final Class<?> cls;
+    
+    /**
+     * TextFromStandardInputStream by default only provides input to the stream
+     * once.
+     */
+    private final Collection<String> inputLines = new ArrayList<>();
+    
+    private boolean inputStarted = false;
+    
+    // Rules that are not treated as rules.
+    private final MessageWriter feedback = new MessageWriter();
+    private final TextFromStandardInputStream stdin = emptyStandardInputStream();
+    private final SystemOutRule stdout = new SystemOutRule().enableLog().mute();
+    
+    /**
+     * Interactive constructor.
+     *
+     * @param cls the class to test
+     */
+    private Interactive(final Class<?> cls) {
+        this.cls = cls;
+    }
+    
+    @Override
+    public Statement apply(final Statement base, final Description description) {
+        // Ugly hack to use existing Rules as Rules.
+        return this.feedback.apply(this.stdin.apply(this.stdout.apply(
+            base, description),
+            description),
+            description);
+    }
+    
+    /**
+     * Calls the .main()-method of the class under test.
+     *
+     * @return fluent
+     * @throws Throwable exception thrown by the program
+     */
+    public Interactive callMain() throws Throwable {
```

Comment:
I understand the rationale to bundle these 3 rules, but I'm not sure about the `callMain`. I think I'd prefer something like

```java
interactive.prepareWithInput("some input, other methods as well");
Echo.main();
Assert.assertEquals("some input, other methods as well", interactive.cleanupWithOutput());
```

Pro: you get the normal compilation errors when the student doesn't provide the correct main method.

Con: it's probably a bit more writing.

Your choice.

Reply:
I decided to leave it as-is. If the student doesn't provide the correct main method, the following message will be returned as an error:

`Method not found: public static void main(String[])`

This message will already be more descriptive than the _scary_ compiler trace, so this should suffice.